pilot/pkg/config/kube/file/controller.go

```diff
@@ -0,0 +1,207 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pilot/pkg/config/kube/crd"
+	"istio.io/istio/pilot/pkg/model"
+	kubecontroller "istio.io/istio/pilot/pkg/serviceregistry/kube/controller"
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/krt"
+	krtfiles "istio.io/istio/pkg/kube/krt/files"
+)
+
+var errUnsupportedOp = fmt.Errorf("unsupported operation: the file config store is a read-only view")
+
+type kindStore struct {
+	collection krt.Collection[config.Config]
+	index      krt.Index[string, config.Config]
+	handlers   []krt.HandlerRegistration
+}
+
+type Controller struct {
+	data    map[config.GroupVersionKind]kindStore
+	schemas collection.Schemas
+	stop    chan struct{}
+}
+
+type ConfigKind struct {
+	*config.Config
+}
+
+func (c ConfigKind) ResourceName() string {
+	if c.Namespace == "" {
+		return c.GroupVersionKind.String() + "/" + c.Name
+	}
+
+	return c.GroupVersionKind.String() + "/" + c.Namespace + c.Name
+}
+
+func (c ConfigKind) Equals(other ConfigKind) bool {
+	return c.Config.Equals(other.Config)
+}
+
+func NewController(
+	fileDir string,
+	domainSuffix string,
+	schemas collection.Schemas,
+	options kubecontroller.Options,
+) (*Controller, error) {
+	stop := make(chan struct{})
+	opts := krt.NewOptionsBuilder(stop, "file-monitor", options.KrtDebugger)
+	watch, err := krtfiles.NewFolderWatch(fileDir, func(b []byte) ([]*config.Config, error) {
+		return parseInputs(b, domainSuffix)
+	}, stop)
+	if err != nil {
+		return nil, err
+	}
+	mainCollection := krtfiles.NewFileCollection(watch, func(c *config.Config) *ConfigKind {
+		return &ConfigKind{c}
+	}, opts.WithName("main")...)
+
+	data := make(map[config.GroupVersionKind]kindStore)
+	for _, s := range schemas.All() {
+		gvk := s.GroupVersionKind()
+		if _, ok := collections.Pilot.FindByGroupVersionKind(gvk); ok {
+			collection := krt.NewCollection(mainCollection, func(ctx krt.HandlerContext, c ConfigKind) *config.Config {
+				if c.GroupVersionKind == gvk {
+					return c.Config
+				}
+
+				return nil
+			}, opts.WithName(gvk.Kind)...)
+
+			data[gvk] = kindStore{
+				collection: collection,
+				index:      krt.NewNamespaceIndex(collection),
+			}
+		}
+	}
+
+	return &Controller{
+		schemas: schemas,
+		stop:    stop,
+		data:    data,
+	}, nil
+}
+
+func (c *Controller) Schemas() collection.Schemas {
+	return c.schemas
+}
+
+func (c *Controller) Get(typ config.GroupVersionKind, name, namespace string) *config.Config {
+	if data, ok := c.data[typ]; ok {
+		if namespace == "" {
```

Comment:
Do we handle file-based config with empty namespace as `"default"` or `""` and do those keys look like `/name` or `"name"`?  (except for un-namespaced objects like GatewayClass etc)

Reply:
hmm this is a bit beyond my knowledge, if this is the case, then I would need to transform the objects when building the collections as to populate the `namespace` correctly, but we would have to check if there are any cluster scoped schemas

Reply:
@stevenctl I checked previous file monitor implementation and it did not handle namespaces in any special way, this means that if the namespace was empty it was stored in an empty namespace inside the delegate store, which means it could only be retreived if fetched by empty namespace. I think this means that it should be safe for us to do the same.